### PR TITLE
validatorcore: add unit tests

### PR DESCRIPTION
**Pull request by ronaudinho**
add unit tests for validatorcore
```
0chain.net/validatorcore/storage/challenge_handler.go:22:	ChallengeHandler		30.2%
0chain.net/validatorcore/storage/context.go:10:			SetupContext			0.0%
0chain.net/validatorcore/storage/handler.go:15:			SetupHandlers			0.0%
0chain.net/validatorcore/storage/handler.go:20:			DumpGoRoutines			0.0%
0chain.net/validatorcore/storage/models.go:42:			String				83.3%
0chain.net/validatorcore/storage/models.go:65:			GetHash				100.0%
0chain.net/validatorcore/storage/models.go:69:			CalculateHash			100.0%
0chain.net/validatorcore/storage/models.go:79:			GetNumBlocks			100.0%
0chain.net/validatorcore/storage/models.go:82:			GetType				100.0%
0chain.net/validatorcore/storage/models.go:97:			GetHashData			100.0%
0chain.net/validatorcore/storage/models.go:112:			GetHash				100.0%
0chain.net/validatorcore/storage/models.go:116:			CalculateHash			100.0%
0chain.net/validatorcore/storage/models.go:120:			GetNumBlocks			100.0%
0chain.net/validatorcore/storage/models.go:123:			GetType				100.0%
0chain.net/validatorcore/storage/models.go:135:			Parse				81.0%
0chain.net/validatorcore/storage/models.go:199:			VerifyBlockNum			90.6%
0chain.net/validatorcore/storage/models.go:249:			VerifyPath			100.0%
0chain.net/validatorcore/storage/models.go:263:			Verify				80.0%
0chain.net/validatorcore/storage/models.go:291:			VerifyChallenge			17.2%
0chain.net/validatorcore/storage/models.go:358:			Sign				100.0%
0chain.net/validatorcore/storage/protocol.go:36:		GetProtocolImpl			0.0%
0chain.net/validatorcore/storage/protocol.go:73:		VerifyAllocationTransaction	0.0%
0chain.net/validatorcore/storage/protocol.go:86:		VerifyChallengeTransaction	0.0%
0chain.net/validatorcore/storage/protocol.go:127:		OnWalletCreateComplete		0.0%
0chain.net/validatorcore/storage/protocol.go:132:		RegisterValidator		0.0%
0chain.net/validatorcore/storage/writemarker/entity.go:27:	GetHashData			100.0%
0chain.net/validatorcore/storage/writemarker/entity.go:34:	VerifySignature			87.5%
0chain.net/validatorcore/storage/writemarker/entity.go:47:	Verify				90.0%
total:								(statements)			49.5%
```

### CODE REVIEW

File: code/go/0chain.net/validatorcore/storage/models.go [hunk 1]
```diff
@@ -207,7 +207,7 @@ func (op *ObjectPath) VerifyBlockNum(challengeRand int64) error {
 	blockNum = blockNum + 1
 
 	if op.RootObject.NumBlocks < blockNum {
-		return common.NewError("invalid_block_num", "Invalid block number"+string(op.RootObject.NumBlocks)+" / "+string(blockNum))
+		return common.NewError("invalid_block_num", "Invalid block number"+fmt.Sprint(op.RootObject.NumBlocks)+" / "+fmt.Sprint(blockNum))
```
Comment (ronaudinho) [inline code/go/0chain.net/validatorcore/storage/models.go:210]: compiler complains, can revert if necessary.

File: code/go/0chain.net/validatorcore/storage/models_test.go [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+	"0chain.net/core/node"
+	"0chain.net/validatorcore/storage"
+
+	"github.com/0chain/gosdk/core/zcncrypto"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+func TestAttributes_String(t *testing.T) {
```
Comment (ronaudinho) [inline code/go/0chain.net/validatorcore/storage/models_test.go:18]: not sure if this is the intended behavior, but `attributes.String()` returns `{}` when owner pays. I would think the check is done elsewhere?

File: code/go/0chain.net/validatorcore/storage/models.go (hunk 1 above)
Comment (peterlimg) [reply code/go/0chain.net/validatorcore/storage/models.go:210]: fmt.Sprintf("Invalid block number %d / %d", op.RootObject.NumBlocks, blockNum) would be better.

File: code/go/0chain.net/validatorcore/storage/models_test.go [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+			allocID: "1",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.objPath.VerifyPath(tt.allocID)
+			if !tt.wantErr {
+				require.NoError(t, err)
+			}
```
Comment (peterlimg) [inline code/go/0chain.net/validatorcore/storage/models_test.go:759]: Errors may also need to check.

File: code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+func TestWriteMarker_GetHashData(t *testing.T) {
+	wm, wallet, err := setupEntityTest(t)
+	require.NoError(t, err)
+
+	want := fmt.Sprintf("%v:%v:%v:%v:%v:%v:%v", "alloc_root", "prev_alloc_root", "alloc_id", "blobber_id", wallet.ClientID, 1, common.Now())
+	got := wm.GetHashData()
+	if got != want {
+		t.Errorf("want %s, got %s", want, got)
+	}
```
Comment (peterlimg) [inline code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go:25]: require.Equal(want, got) might be better.

File: code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+			publicKey: "1",
+			wantErr:   true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := wm.Verify(tt.allocID, tt.allocRoot, tt.publicKey)
+			if !tt.wantErr {
+				require.NoError(t, err)
+			}
```
Comment (peterlimg) [inline code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go:101]: The same, check errors

File: code/go/0chain.net/validatorcore/storage/models_test.go [hunk 6]
```diff
@@ -0,0 +1,879 @@
...
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.chReq.VerifyChallenge(tt.ch, tt.alloc)
+			if !tt.wantErr {
+				require.NoError(t, err)
+			}
```
Comment (peterlimg) [inline code/go/0chain.net/validatorcore/storage/models_test.go:842]: Check errors

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/validatorcore/storage/models.go (hunk 1 above)
Comment (ronaudinho) [reply code/go/0chain.net/validatorcore/storage/models.go]: updated

File: code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go (hunk 5 above)
Comment (ronaudinho) [reply code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go:104]: updated

File: code/go/0chain.net/validatorcore/storage/models_test.go (hunk 6 above)
Comment (ronaudinho) [reply code/go/0chain.net/validatorcore/storage/models_test.go:716]: updated

File: code/go/0chain.net/validatorcore/storage/models_test.go (hunk 3 above)
Comment (ronaudinho) [reply code/go/0chain.net/validatorcore/storage/models_test.go:627]: updated

File: code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go (hunk 4 above)
Comment (ronaudinho) [reply code/go/0chain.net/validatorcore/storage/writemarker/entity_test.go]: updated

**Review by peterlimg — APPROVED** [review]
👍

**Pull request merged**